### PR TITLE
[POAE7-2341] add try-cast stringOp to support cast to date from non-encoded string

### PR DESCRIPTION
**Pull request by spevenhe**
### What changes were proposed in this pull request?

1. add try-cast stringOp to support cast to date from non-encoded string
2. add date format string generating patterns for test
### Why are the changes needed?

to support tpch for velox
### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

Uts

### CODE REVIEW

**Review by chaojun-zhang — COMMENTED** [review]
LGTM for the substrait json file

File: cider/function/string/StringOps.h [hunk 1]
```diff
@@ -44,7 +44,7 @@ namespace StringOps_Namespace {
 
 struct NullableStrType {
   NullableStrType(const std::string& str) : str(str), is_null(false) {}
-  NullableStrType(const std::string_view& sv) : str(sv), is_null(false) {}
+  NullableStrType(const std::string_view sv) : str(sv), is_null(false) {}
```
Comment (jikunshang) [inline cider/function/string/StringOps.h:47]: should remove reference here?

File: cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp [hunk 2]
```diff
@@ -970,6 +970,36 @@ std::shared_ptr<Analyzer::Expr> Substrait2AnalyzerExprConverter::toAnalyzerExpr(
...
+        auto iter = expr_map_ptr->find(col_id);
+        if (iter != expr_map_ptr->end()) {
+          auto type = iter->second->get_type_info().get_type();
+          if (type == SQLTypes::kTEXT || type == SQLTypes::kVARCHAR) {
+            std::vector<std::shared_ptr<Analyzer::Expr>> args;
+            args.push_back(toAnalyzerExpr(
+                s_cast_expr.input().selection(), function_map, expr_map_ptr));
+
+            return makeExpr<Analyzer::TryStringCastOper>(SQLTypes::kDATE, args);
+            ;
```
Comment (jikunshang) [inline cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp:992]: remove.

File: cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp [hunk 3]
```diff
@@ -970,6 +970,36 @@ std::shared_ptr<Analyzer::Expr> Substrait2AnalyzerExprConverter::toAnalyzerExpr(
     // use a int32/int64 to represent under substrait protocol
     if (!s_cast_expr.input().literal().fixed_char().empty()) {
       v.bigintval = dateToInt64(s_cast_expr.input().literal().fixed_char());
+    }
+    // TODO: spevenhe.
+    // support nested function like cast(cast(xx as String) AS Date).
+    // CAST(SUBSTRING(col_string ,0,5) AS DATE
+    // Now only supports cast(col_string AS Date)
+    else if (s_cast_expr.input().selection().has_direct_reference() &&
```
Comment (jikunshang) [inline cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp:978]: can you refactor `selection().has_direct_reference()` and `selection().has_root_reference()` as a new function, I guess  this is checking whether expr is a column.

**Review by yma11 — DISMISSED** [review]

File: cider/function/string/StringOps.h (hunk 1 above)
Comment (spevenhe) [reply cider/function/string/StringOps.h:47]: std::string_view is a read-only view over a string, it only holds a pointer to the first character of the string and its length. It says the best practice to pass string_view is by value

File: cider/util/Datum.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  bool is_null;
+
+  DEVICE VarlenDatum() : length(0), pointer(nullptr), is_null(true) {}
+  DEVICE virtual ~VarlenDatum() {}
+
+  VarlenDatum(const size_t l, int8_t* p, const bool n)
+      : length(l), pointer(p), is_null(n) {}
+};
+
+union Datum {
```
[the remark this answers is not in the document]
Comment (spevenhe) [reply cider/util/Datum.h:49]: It is based on the latest heavydb code. And I thought putting datum in datum.h makes sense.

File: cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp [hunk 5]
```diff
@@ -970,6 +970,36 @@ std::shared_ptr<Analyzer::Expr> Substrait2AnalyzerExprConverter::toAnalyzerExpr(
...
+        } else {
+          throw std::runtime_error("Failed to get field reference expr.");
+        }
+      }
+      throw std::runtime_error(
+          "Unsupported date cast: can not get the origin type of cast");
+
     } else {
       throw std::runtime_error("Not supported literal type, id: " +
                                s_cast_expr.input().literal().literal_type_case());
```
Comment (kd97) [inline cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp:1005]: What if cast Timestamp to Date, now will go into this branch? need to refine this error message clearly?

Comment (kd97) [reply cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp:1005]: also, encapsulate this part logic to reduce some if-else branches better？

Comment (spevenhe) [reply cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp:1005]: I think for timestamp, we can just add ```type == SQLTypes::kTIMESTAMP``` in ``` if (type == SQLTypes::kTEXT || type == SQLTypes::kVARCHAR)```.  
The error message seems not clear, let me refine.

File: cider/util/Datum.cpp [hunk 6]
```diff
@@ -263,12 +263,73 @@ T parseInteger(std::string_view s, SQLTypeInfo const& ti) {
...
+      break;
+    // case kPOINT:
+    // case kMULTIPOINT:
+    // case kLINESTRING:
+    // case kMULTILINESTRING:
+    // case kPOLYGON:
+    // case kMULTIPOLYGON:
+    //   throw std::runtime_error("Internal error: geometry type in NullDatum.");
+    default:
+      throw std::runtime_error("Internal error: invalid type in NullDatum.");
```
[the remark this answers is not in the document]
Comment (spevenhe) [reply cider/util/Datum.cpp:312]: updated

File: cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp (hunk 3 above)
Comment (spevenhe) [reply cider/exec/plan/parser/SubstraitToAnalyzerExpr.cpp]: Yes, it is. Updated

File: cider/util/Datum.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ * @file		Datum.h
+ * @brief	 Definitions for core Datum union type
+ *
+ */
+
+#pragma once
+
+#include "type/data/funcannotations.h"
+
+#ifndef __CUDACC__
```
Comment (jikunshang) [inline cider/util/Datum.h:33]: Please remove this.

**Review by kd97 — APPROVED** [review]

**Review by jikunshang — COMMENTED** [review]
Overall LGTM.

**Review by jikunshang — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Review by yma11 — APPROVED** [review]

File: cider/function/string/StringOps.cpp [hunk 8]
```diff
@@ -78,6 +78,24 @@ std::regex StringOp::generateRegex(const std::string& op_name,
...
+}
+
+Datum TryStringCast::numericEval(const std::string_view str) const {
+  if (str.empty()) {
+    return NullDatum(return_ti_);
+  }
+  // Need to make copy for now b/c StringToDatum can mod SQLTypeInfo arg
+  SQLTypeInfo return_ti(return_ti_);
+  try {
+    return StringToDatum(str, return_ti);
```
Comment (kd97) [inline cider/function/string/StringOps.cpp:93]: can we skip cast_type branch check inside StringToDatum method? eg.separate and call different cast function from here string_to_int, string_to_date...

File: cider/exec/template/StringOpsIR.cpp [hunk 9]
```diff
@@ -111,6 +111,27 @@ extern "C" RUNTIME_EXPORT int32_t lower_encoded(int32_t string_id,
...
 
+#define DEF_APPLY_NUMERIC_STRING_OPS(value_type, value_name)                             \
+  extern "C" RUNTIME_EXPORT ALWAYS_INLINE value_type                                     \
+      apply_numeric_string_ops_##value_name(                                             \
+          const char* str_ptr, const int32_t str_len, const int64_t string_ops_handle) { \
+    std::string raw_str(str_ptr, str_len);                                               \
+    auto string_ops =                                                                    \
+        reinterpret_cast<const StringOps_Namespace::StringOps*>(string_ops_handle);      \
+    const auto result_datum = string_ops->numericEval(raw_str);                          \
+    return result_datum.value_name##val;                                                 \
```
Comment (kd97) [inline cider/exec/template/StringOpsIR.cpp:122]: can we skip cast_type branch check inside` numericEval` method? eg.separate and call different cast function from here string_to_int, string_to_date...

Comment (spevenhe) [reply cider/exec/template/StringOpsIR.cpp:122]: I think we can. It is a good place to switch type and cast string to there different types

**Pull request merged**